### ISSUE TITLE
TiledRange serialization problem with Cereal text archives 

### ISSUE THREAD

**Issue by keceli** [ValeevGroup/tiledarray#345]
When I try to serialize a `TiledRange` object into a Cereal JSON or XML archive, I get a compilation error. Cereal binary archive or madness text archive work fine. Here is the error:
```
In file included from /home/keceli/nwx/tiledarray/src/TiledArray/tensor/utility.h:33,
                 from /home/keceli/nwx/tiledarray/src/TiledArray/tensor/kernels.h:30,
                 from /home/keceli/nwx/tiledarray/src/TiledArray/tensor/tensor.h:28,
                 from /home/keceli/nwx/tiledarray/src/TiledArray/tensor.h:31,
                 from /home/keceli/nwx/tiledarray/src/tiledarray.h:29,
                 from /home/keceli/nwx/tiledarray/examples/demo/demo.cpp:25:
/home/keceli/nwx/tiledarray/src/TiledArray/tiled_range1.h: In instantiation of ‘void TiledArray::TiledRange1::serialize(const Archive&) const [with Archive = cereal::JSONOutputArchive; typename std::enable_if<is_output_archive_v<Archive> >::type* <anonymous> = 0]’:
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/cereal-src/include/cereal/access.hpp:246:51:   required from ‘static decltype (t.serialize(ar)) cereal::access::member_serialize(Archive&, T&) [with Archive = cereal::JSONOutputArchive; T = TiledArray::TiledRange1; decltype (t.serialize(ar)) = void]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/cereal-src/include/cereal/cereal.hpp:481:33:   required from ‘ArchiveType& cereal::OutputArchive<ArchiveType, Flags>::processImpl(const T&) [with T = TiledArray::TiledRange1; typename cereal::traits::detail::EnableIfHelper<cereal::traits::has_member_serialize<T, ArchiveType>::value, (! cereal::traits::has_invalid_output_versioning<T, ArchiveType>::value), (cereal::traits::is_output_serializable<T, ArchiveType>::value && (cereal::traits::is_specialized_member_serialize<T, ArchiveType>::value || (! cereal::traits::is_specialized<T, ArchiveType>::value)))>::type <anonymous> = (cereal::traits::detail::sfinae)0; ArchiveType = cereal::JSONOutputArchive; unsigned int Flags = 0]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/cereal-src/include/cereal/cereal.hpp:417:9:   required from ‘void cereal::OutputArchive<ArchiveType, Flags>::process(T&&) [with T = const TiledArray::TiledRange1&; ArchiveType = cereal::JSONOutputArchive; unsigned int Flags = 0]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/cereal-src/include/cereal/cereal.hpp:311:9:   required from ‘ArchiveType& cereal::OutputArchive<ArchiveType, Flags>::operator()(Types&& ...) [with Types = {const TiledArray::TiledRange1&}; ArchiveType = cereal::JSONOutputArchive; unsigned int Flags = 0]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/madness-src/src/madness/world/cereal_archive.h:83:16:   required from ‘std::enable_if_t<((! madness::is_trivially_serializable<T>::value) || cereal::traits::is_text_archive<Cereal>::value), void> madness::archive::CerealOutputArchive<Muesli>::store(const T*, long int) const [with T = TiledArray::TiledRange1; Cereal = cereal::JSONOutputArchive; Muesli = cereal::JSONOutputArchive; std::enable_if_t<((! madness::is_trivially_serializable<T>::value) || cereal::traits::is_text_archive<Cereal>::value), void> = void]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/madness-src/src/madness/world/archive.h:433:13:   [ skipping 9 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/madness-src/src/madness/world/archive.h:814:54:   required from ‘std::enable_if_t<is_output_archive_v<Archive>, const Archive&> madness::archive::operator&(const Archive&, const T&) [with Archive = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; T = TiledArray::TiledRange; std::enable_if_t<is_output_archive_v<Archive>, const Archive&> = const madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>&]’
/home/keceli/nwx/tiledarray/src/TiledArray/conversions/../dist_array.h:1283:69:   required from ‘void TiledArray::DistArray< <template-parameter-1-1>, <template-parameter-1-2> >::serialize(const Archive&) const [with Archive = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; <template-parameter-2-2> = void; Tile = TiledArray::Tensor<double>; Policy = TiledArray::DensePolicy]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/madness-src/src/madness/world/archive.h:621:19:   required from ‘static std::enable_if_t<((is_output_archive_v<A> && (! std::is_function<U>::value)) && (((has_member_serialize_v<U, A> || has_nonmember_serialize_v<U, A>) || has_freestanding_serialize_v<U, A>) || has_freestanding_default_serialize_v<U, A>)), void> madness::archive::ArchiveStoreImpl<Archive, T, Enabler>::store(const A&, const U&) [with A = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; U = TiledArray::DistArray<TiledArray::Tensor<double>, TiledArray::DensePolicy>; Archive = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; T = TiledArray::DistArray<TiledArray::Tensor<double>, TiledArray::DensePolicy>; Enabler = void; std::enable_if_t<((is_output_archive_v<A> && (! std::is_function<U>::value)) && (((has_member_serialize_v<U, A> || has_nonmember_serialize_v<U, A>) || has_freestanding_serialize_v<U, A>) || has_freestanding_default_serialize_v<U, A>)), void> = void]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/madness-src/src/madness/world/archive.h:733:51:   required from ‘static const Archive& madness::archive::ArchiveImpl<Archive, T, Enabler>::wrap_store(const Archive&, const T&) [with A = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; <template-parameter-2-2> = void; Archive = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; T = TiledArray::DistArray<TiledArray::Tensor<double>, TiledArray::DensePolicy>; Enabler = void]’
/home/keceli/nwx/tiledarray/build_ta_220531/_deps/madness-src/src/madness/world/archive.h:772:54:   required from ‘std::enable_if_t<is_output_archive_v<Archive>, const Archive&> madness::archive::operator<<(const Archive&, const T&) [with Archive = madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>; T = TiledArray::DistArray<TiledArray::Tensor<double>, TiledArray::DensePolicy>; std::enable_if_t<is_output_archive_v<Archive>, const Archive&> = const madness::archive::CerealOutputArchive<cereal::JSONOutputArchive>&]’
/home/keceli/nwx/tiledarray/examples/demo/demo.cpp:122:62:   required from here
/home/keceli/nwx/tiledarray/src/TiledArray/tiled_range1.h:219:7: error: no match for ‘operator&’ (operand types are ‘const cereal::JSONOutputArchive’ and ‘const range_type’ {aka ‘const std::pair<long int, long int>’})
  219 |     ar& range_& elements_range_& tiles_ranges_& elem2tile_;
      |     ~~^~~~~~~~
compilation terminated due to -Wfatal-errors.
```

While the error points to `TiledRange1` serialization, I didn't see a problem in serializing `TiledRange1` objects. The compilation fails only when I try to serialize `TiledRange` objects or any tensor types into Cereal text archives. Probably related to cereal/madness traits, but I couldn't figure out.

One can reproduce this error by adding the following to the `demo.cpp` example.

```
#include <cereal/archives/json.hpp>
#include <madness/world/archive.h>
```
...
```
std::stringstream ss;
  {
    madness::archive::CerealOutputArchive<cereal::JSONOutputArchive> arout(ss);
    arout <<  TR;
  }
```

**Comment by evaleev** [comment]
the issue is 2-fold:
- `serialize()` calls break const-correctness (unnoticed because MADNESS serialization itself breaks const-correctness ... and because I suspect MADNESS-Cereal interface also breaks it) ... this is addressed by #349
- JSON serialization requires appropriate Cereal headers to be included, e.g. a `cereal/types/utility.hpp` and `cereal/types/vector.hpp`

**Comment by keceli** [comment]
Thank you @evaleev for the solution. It works now.
